test_scripts/clientconf.go

```diff
@@ -0,0 +1,182 @@
+	decoys := clientConf.DecoyList.TlsDecoys
+	fmt.Printf("Decoy List: %d decoys\n", len(decoys))
+	for i, decoy := range decoys {
+		ip := make(net.IP, 4)
+		binary.BigEndian.PutUint32(ip, decoy.GetIpv4Addr())
+		fmt.Printf("%d:\n  %s (%s)\n", i, decoy.GetHostname(), ip.To4().String())
+		if decoy.GetPubkey() != nil {
+			fmt.Printf("  pubkey: %s\n", hex.EncodeToString(decoy.GetPubkey().Key[:]))
+		}
+		if decoy.GetTimeout() != 0 {
+			fmt.Printf("  timeout: %d ms\n", decoy.GetTimeout())
+		}
+		if decoy.GetTcpwin() != 0 {
+			fmt.Printf("  tcpwin: %d bytes\n", decoy.GetTcpwin())
+		}
+	}
+
+}
+
+func parseClientConf(fname string) pb.ClientConf {
+
+	clientConf := pb.ClientConf{}
+	buf, err := ioutil.ReadFile(fname)
+	if err != nil {
+		log.Fatal("Error reading file:", err)
+	}
+	err = proto.Unmarshal(buf, &clientConf)
+	if err != nil {
+		log.Fatal("Error parsing ClientConf", err)
+	}
+	return clientConf
+}
+
+func parsePubkey(pubkey string) []byte {
+	pubkey_bin, err := hex.DecodeString(pubkey)
+	if err != nil {
+		log.Fatal("Error parsing pubkey:", err)
+	}
+	if len(pubkey_bin) != 32 {
+		log.Fatal("Error: pubkey length: expected 32, got ", len(pubkey_bin))
+	}
+	return pubkey_bin
+
+}
+
+func updateDecoy(decoy *pb.TLSDecoySpec, host string, ip string, pubkey string, delpubkey bool, timeout int, tcpwin int) {
+
+	if host != "" {
+		decoy.Hostname = &host
+	}
+	if ip != "" {
+		ip4 := binary.BigEndian.Uint32(net.ParseIP(ip).To4())
+		decoy.Ipv4Addr = &ip4
+	}
+	if pubkey != "" {
+		decoy.Pubkey.Key = parsePubkey(pubkey)
+	}
+	if delpubkey {
+		decoy.Pubkey = nil
+	}
+	if timeout != 0 {
+		t := uint32(timeout)
+		decoy.Timeout = &t
+	}
+	if tcpwin != 0 {
+		t := uint32(tcpwin)
+		decoy.Tcpwin = &t
+	}
+}
+
+func main() {
+	var fname = flag.String("f", "", "`ClientConf` file to parse")
+	var out_fname = flag.String("o", "", "`output` file name to write new/modified config")
+	var generation = flag.Int("generation", 0, "New/modified generation")
+	var pubkey = flag.String("pubkey", "", "New/modified (decoy) pubkey. If -add or -update, applies to specific decoy. If -all applies to all decoys. Otherwise, applies to default pubkey.")
+	var delpubkey = flag.Bool("delpubkey", false, "Delete pubkey from decoy (or all if -all)")
+
+	var add = flag.Bool("add", false, "Add a new decoy. Must set -host and -ip, optionally set -timeout, -tcpwin and -pubkey")
+	var delete = flag.Int("delete", -1, "Delete a decoy at `index`")
+	var update = flag.Int("update", -1, "Update a decoy at `index`")
+
+	var host = flag.String("host", "", "New/modified decoy host")
+	var ip = flag.String("ip", "", "New/modified IP address")
+	var timeout = flag.Int("timeout", 0, "New/modified timeout")
+	var tcpwin = flag.Int("tcpwin", 0, "New/modified tcpwin")
+
+	var all = flag.Bool("all", false, "If set, replace all pubkeys/timeouts/tcpwins in decoy list with pubkey/timeout/tcpwin if provided")
```

Comment:
Says "replace all pubkeys/timeouts/tcpwins", but will replace all entries with decoy, host and ip, if those are provided.

Reply:
by design, though could be better documented. perhaps one wants to replace all the decoys with a test decoy IP in a ClientConf, for example.

Reply:
> perhaps one wants to replace all the decoys with a test decoy IP in a ClientConf, for example.

perhaps, this behavior is worth documenting, especially if such a use-case is supported :)